### ISSUE TITLE
Plans on providing the published results?

### ISSUE THREAD

**Issue by amirjamez** [google/ml-compiler-opt#5]
Hi @mtrofin, 

I was wondering if you planned to publish the results on arXiv or on a conference, if you haven't done so. I recall you mentioned this in the LLVM-Dev. 

Thanks,
- Amir

**Comment by mtrofin** [comment]
Yes - we're slowly getting there. If you are interested in trying it out or
extending/applying to other optimizations, the demo under
github.com/google/ml-compiler-opt, and the llvm code, are probably a better
starting point (happy to detail more).

If you're interested in citing the work, what's your timeframe? - so we can
prioritize putting this on arxiv (otherwise we keep deprioritizing it)

Thanks,
Mircea.

**Comment by amirjamez** [comment]
Thanks. We will be interested in citing the work sometime soon. It would be great to have an official paper to refer to for numbers and methodology.

Sure. I'd appreciate if you provide details on training the RL. Believe should be [this file](https://github.com/google/ml-compiler-opt/blob/main/compiler_opt/rl/trainer_test.py) ?

Regards,
-- Amir

**Comment by mtrofin** [comment]
An end-to-end demo - i.e. how to collect a corpus, how to train, etc - is given [here](https://github.com/google/ml-compiler-opt/blob/main/docs/demo/demo.md). The training driver [is this one](https://github.com/google/ml-compiler-opt/blob/main/compiler_opt/rl/train_locally.py).

**Comment by yundiqian** [comment]
Hi Amir,

Our to-be-published paper might be the best resources for RL details. We are prioritizing it and will publish it hopefully next week. Before that, the end-to-end [demo](https://github.com/google/ml-compiler-opt/blob/main/docs/demo/demo.md) Mircea mentioned may be the best resources to look at

**Comment by mtrofin** [comment]
Re. TF version, it was about a TFAgents dependency. @ebrevdo @yundiqian @jaingaurav for TFAgents details.

**Comment by mtrofin** [comment]
https://arxiv.org/abs/2101.04808 for the paper.

**Comment by yundiqian** [comment]
For TF and TF-Agents versions, we just updated it to the latest. There is no specific feature we need from the latest version, so an earlier version may also work.

**Comment by amirjamez** [comment]
@mtrofin Thanks for the paper.

Looking at the demo tutorial, I don't seem to understand the role of Fushsia in training RL. Can the training driver be used without the Fuchsia OS?

**Comment by mtrofin** [comment]
Fuchsia is just used as a project to extract a corpus from. You can replace it with anything else that's clang-based, so you can extract a compilation database, from which you then extract the IR corpus. We liked Fuchsia because it yields a reasonably sized IR corpus, and it's open source; and we've been working with them because they were interested in size reduction, so a good part of the demo was basically ready-written.

Let us know if you run into any trouble trying it out with a clang-based project.

**Comment by amirjamez** [comment]
Thanks. One small thing is `compiler_opt/tools/generate_tfrecord.py`, has been renamed to `generate_default_trace.py`, you might wanna revise the demo.md as well.

**Comment by mtrofin** [comment]
Thanks - fixed!

**Comment by TH3CHARLie** [comment]
> Fuchsia is just used as a project to extract a corpus from. You can replace it with anything else that's clang-based, so you can extract a compilation database, from which you then extract the IR corpus. We liked Fuchsia because it yields a reasonably sized IR corpus, and it's open source; and we've been working with them because they were interested in size reduction, so a good part of the demo was basically ready-written.
> 
> We'll clarify that in the demo, reading it again, I see that Fuchsia is sprung up rather surprisingly.
> 
> Let us know if you run into any trouble trying it out with a clang-based project.

Hi, thanks for the demo and the great paper! I am a liitle bit confused about what you mean by "clang-based`, is that this method cannot be applied to standard C++ codebase?

**Comment by mtrofin** [comment]
"clang-based" refers to the target project needing to be compiled with clang. That's not a comment about C++ standard compliance, just to the practical realities of C++ projects - switching compilers (e.g. gcc -> clang) is non-trivial, so the comment underlines the fact that, as long as a project is already using clang, one should be able to use this work.

For more context: our work is about replacing the internals of the compiler (starting with inlining decision heuristics) with a learned policy. The work is done in the LLVM (hence clang) repo (llvm-project/llvm). This repo here (google/ml-compiler-opt) handles the training side, which happens offline, outside of the compiler. In addition, our goal is productization. For this latter reason, we're keeping our focus to LLVM, which may impact assumptions in this (google/ml-compiler-opt) repo.

Probably the ideas can be applied elsewhere, and likely those assumptions I mentioned aren't many and the work in this repo can be generalized to handle training data from other compilers, but that's not within the scope of our work. Having said that, if there's strong interest to extend this project, we'd be happy to set it up for contribution and welcome patches!

**Comment by TH3CHARLie** [comment]
Thanks for the clarification! It would be also interesting to see if the model is learnt from

**Comment by mtrofin** [comment]
You mean the corpus the model is trained from? The one we checked in upstream llvm was trained on a Google - internal codebase; that's why we've detailed collection of a corpus from an open-source code base, such as Fuchsia (a probably larger one may be Chromium)

**Comment by TH3CHARLie** [comment]
Sorry, my comment was some kind of draft but it got sent out somehow. But
yes my question was about the corpus. I am thinking about applying this
year's LLVM GSoC on using learning heuristics to do loop transforms so this
paper as well as the demo is super useful.

Mircea Trofin ***@***.***> 于2021年3月19日周五 上午1:32写道：

**Comment by mtrofin** [comment]
Sorry, I saw the comment go, so I deleted mine, too, figuring maybe there was a mistake :) 

Looking forward to collaborating on learned loop transforms, should you decide to apply to GSoC - let us know if you need more info meanwhile, too!

Also, there's this monthly meeting (described [here](https://docs.google.com/document/d/1JecbplF09l3swTjze-UVeLh4L48svJxGVy4mz_e9Rhs/edit#)) - feel free to join! Perhaps a reasonable topic for the agenda for the April meeting should be the related GSoC projects.

**Comment by TH3CHARLie** [comment]
Thanks! This would be super useful and I am looking forward to next month's meeting. I am already drafting my purpose for the project and it would be great if you can take a look. Any feedbacks would be helpful!
